[FSDP][state_dict][1/N] Moving state_dict logic to pre_state_dict_hook

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88638
* #88637
* #88636
* #88635
* #88481
* __->__ #87900

This is one step toward the ultimate goal: remove the overwritten state_dict in FSDP. All the logic should be either in `pre_state_dict_hook` or `post_state_dict_hook`. 

Since current `nn.Module` does not support `pre_state_dict_hook`, this PR mimic `pre_state_dict_hook` by calling the pre hook inside post the hook, effectively ditching all the work done by `nn.Module.state_dict`. Once `pre_state_dict_hook` is supported by `nn.Module`, these pre hook calls can be moved out from the post hooks and be registered to `nn.Module.pre_state_dict_hook`.

The major issue of this temporary solution is that `post_state_dict_hook` is called from the leaf node to the root node. This makes the `module._lazy_init()` invalid as FSDP assumes `_lazy_init()` to be called from the root. As a result, `FSDP.state_dict` currently contains only one logic -- calling `module._lazy_init()`.

fegin committed:
[FSDP] Moving state_dict logic to pre_state_dict_hook

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Moving state_dict logic to pre_state_dict_hook"




[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Moving state_dict logic to pre_state_dict_hook"


This is one step toward the ultimate goal: remove the overwritten state_dict in FSDP. All the logic should be either in `pre_state_dict_hook` or `post_state_dict_hook`. 

Since current `nn.Module` does not support `pre_state_dict_hook`, this PR mimic `pre_state_dict_hook` by calling the pre hook inside post the hook, effectively ditching all the work done by `nn.Module.state_dict`. Once `pre_state_dict_hook` is supported by `nn.Module`, these pre hook calls can be moved out from the post hooks and be registered to `nn.Module.pre_state_dict_hook`.

The major issue of this temporary solution is that `post_state_dict_hook` is called from the leaf node to the root node. This makes the `module._lazy_init()` invalid as FSDP assumes `_lazy_init()` to be called from the root. As a result, `FSDP.state_dict` currently contains only one logic -- calling `module._lazy_init()`.


[ghstack-poisoned]

zhaojuanmao reviewed (commented):
awesome! overall look great to me, will let Rohan or Andrew accept

rohan-varma commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -345,7 +345,7 @@ def test_state_dict_rank0_offload_save_load_flow(self, use_orig_params: bool):
     @parametrize("fp16", [True, False])
     @parametrize("state_dict_rank0_and_offload", [True, False])
-    @parametrize("use_orig_params", [False, True])
+    @parametrize("use_orig_params", [True, False])
```
nit: why do we need to change this order?

fegin replied:
uh, this is just to make this line be consistent with the previous lines.

rohan-varma commented on torch/distributed/fsdp/_state_dict_utils.py:
```diff
@@ -26,48 +27,217 @@
+    rank0_only: bool,
+) -> None:
+    """
+    Performs the pre-state_dict tasks sharded by all state_dict types that require
```
nit: "sharded" -> shared?

rohan-varma commented on torch/distributed/fsdp/_state_dict_utils.py:
```diff
@@ -26,48 +27,217 @@
+        param = functools.reduce(getattr, fqn.split("."), module.module)
         fqn = f"{prefix}{fqn}"
+        if no_fsdp_return:
+            state_dict.pop(fqn)
```
is this to remove FSDP managed parameters from the state_dict if we are rank0_only and not rank 0?

rohan-varma replied:
And would we still need this once we have the proper support for pre_state_dict hook in nn.Module?

fegin replied:
1. Yes
2. IMO, we still need this line. I think the original rank_only implementation may not be correct. The original implementation simply set state_dict to {} if `rank_only and rank !=0`. However, this may be a problem if ignore parameters are set. In such a case, we should not simply throw away what are inside `state_dict` on other ranks.

awgu replied:
> IMO, we still need this line. I think the original rank_only implementation may not be correct. The original implementation simply set state_dict to {} if rank_only and rank !=0. However, this may be a problem if ignore parameters are set. In such a case, we should not simply throw away what are inside state_dict on other ranks.

Good call!

rohan-varma commented on torch/distributed/fsdp/_state_dict_utils.py:
```diff
@@ -26,48 +27,217 @@
+        if no_fsdp_return:
+            state_dict.pop(fqn)
+            continue
+        state_dict[fqn] = param
```
is this where we're actually populating the state_dict? i.e. we're not relying on anything that the nn.Module state_dict implementation does anymore?

fegin replied:
Yes.

rohan-varma reviewed (approved):
LGTM overall, but agree this is a somewhat hacky solution because we're calling the state_dict, then in post-hook processing essentially implementing our own version of state_dict by running a pre hook, discarding super().state_dict results, and then doing post-hook. 

Adding nn.Module pre state_dict hook immediately is probably better solution, but might take ~2-3 weeks. If that timeline doesn't work then we can go ahead and land this change.

fegin committed:
Update on "[FSDP] Moving state_dict logic to pre_state_dict_hook"


This is one step toward the ultimate goal: remove the overwritten state_dict in FSDP. All the logic should be either in `pre_state_dict_hook` or `post_state_dict_hook`. 

Since current `nn.Module` does not support `pre_state_dict_hook`, this PR mimic `pre_state_dict_hook` by calling the pre hook inside post the hook, effectively ditching all the work done by `nn.Module.state_dict`. Once `pre_state_dict_hook` is supported by `nn.Module`, these pre hook calls can be moved out from the post hooks and be registered to `nn.Module.pre_state_dict_hook`.

The major issue of this temporary solution is that `post_state_dict_hook` is called from the leaf node to the root node. This makes the `module._lazy_init()` invalid as FSDP assumes `_lazy_init()` to be called from the root. As a result, `FSDP.state_dict` currently contains only one logic -- calling `module._lazy_init()`.


[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Moving state_dict logic to pre_state_dict_hook"


This is one step toward the ultimate goal: remove the overwritten state_dict in FSDP. All the logic should be either in `pre_state_dict_hook` or `post_state_dict_hook`. 

Since current `nn.Module` does not support `pre_state_dict_hook`, this PR mimic `pre_state_dict_hook` by calling the pre hook inside post the hook, effectively ditching all the work done by `nn.Module.state_dict`. Once `pre_state_dict_hook` is supported by `nn.Module`, these pre hook calls can be moved out from the post hooks and be registered to `nn.Module.pre_state_dict_hook`.

The major issue of this temporary solution is that `post_state_dict_hook` is called from the leaf node to the root node. This makes the `module._lazy_init()` invalid as FSDP assumes `_lazy_init()` to be called from the root. As a result, `FSDP.state_dict` currently contains only one logic -- calling `module._lazy_init()`.


[ghstack-poisoned]

fegin committed:
Update on "[FSDP][state_dict][1/N] Moving state_dict logic to pre_state_dict_hook"


This is one step toward the ultimate goal: remove the overwritten state_dict in FSDP. All the logic should be either in `pre_state_dict_hook` or `post_state_dict_hook`. 

Since current `nn.Module` does not support `pre_state_dict_hook`, this PR mimic `pre_state_dict_hook` by calling the pre hook inside post the hook, effectively ditching all the work done by `nn.Module.state_dict`. Once `pre_state_dict_hook` is supported by `nn.Module`, these pre hook calls can be moved out from the post hooks and be registered to `nn.Module.pre_state_dict_hook`.

The major issue of this temporary solution is that `post_state_dict_hook` is called from the leaf node to the root node. This makes the `module._lazy_init()` invalid as FSDP assumes `_lazy_init()` to be called from the root. As a result, `FSDP.state_dict` currently contains only one logic -- calling `module._lazy_init()`.


[ghstack-poisoned]

fegin committed:
Update on "[FSDP][state_dict][1/N] Moving state_dict logic to pre_state_dict_hook"


This is one step toward the ultimate goal: remove the overwritten state_dict in FSDP. All the logic should be either in `pre_state_dict_hook` or `post_state_dict_hook`. 

Since current `nn.Module` does not support `pre_state_dict_hook`, this PR mimic `pre_state_dict_hook` by calling the pre hook inside post the hook, effectively ditching all the work done by `nn.Module.state_dict`. Once `pre_state_dict_hook` is supported by `nn.Module`, these pre hook calls can be moved out from the post hooks and be registered to `nn.Module.pre_state_dict_hook`.

The major issue of this temporary solution is that `post_state_dict_hook` is called from the leaf node to the root node. This makes the `module._lazy_init()` invalid as FSDP assumes `_lazy_init()` to be called from the root. As a result, `FSDP.state_dict` currently contains only one logic -- calling `module._lazy_init()`.


[ghstack-poisoned]

fegin committed:
Update on "[FSDP][state_dict][1/N] Moving state_dict logic to pre_state_dict_hook"


This is one step toward the ultimate goal: remove the overwritten state_dict in FSDP. All the logic should be either in `pre_state_dict_hook` or `post_state_dict_hook`. 

Since current `nn.Module` does not support `pre_state_dict_hook`, this PR mimic `pre_state_dict_hook` by calling the pre hook inside post the hook, effectively ditching all the work done by `nn.Module.state_dict`. Once `pre_state_dict_hook` is supported by `nn.Module`, these pre hook calls can be moved out from the post hooks and be registered to `nn.Module.pre_state_dict_hook`.

The major issue of this temporary solution is that `post_state_dict_hook` is called from the leaf node to the root node. This makes the `module._lazy_init()` invalid as FSDP assumes `_lazy_init()` to be called from the root. As a result, `FSDP.state_dict` currently contains only one logic -- calling `module._lazy_init()`.


[ghstack-poisoned]

fegin commented:
@pytorchbot merge

fegin committed:
Update on "[FSDP][state_dict][1/N] Moving state_dict logic to pre_state_dict_hook"


This is one step toward the ultimate goal: remove the overwritten state_dict in FSDP. All the logic should be either in `pre_state_dict_hook` or `post_state_dict_hook`. 

Since current `nn.Module` does not support `pre_state_dict_hook`, this PR mimic `pre_state_dict_hook` by calling the pre hook inside post the hook, effectively ditching all the work done by `nn.Module.state_dict`. Once `pre_state_dict_hook` is supported by `nn.Module`, these pre hook calls can be moved out from the post hooks and be registered to `nn.Module.pre_state_dict_hook`.

The major issue of this temporary solution is that `post_state_dict_hook` is called from the leaf node to the root node. This makes the `module._lazy_init()` invalid as FSDP assumes `_lazy_init()` to be called from the root. As a result, `FSDP.state_dict` currently contains only one logic -- calling `module._lazy_init()`.


[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_state_dict_utils.py:
```diff
@@ -31,49 +38,218 @@
+    # TODO: change to this call after pre_state_dict_hook is in `nn.Module`.
+    # if module.is_root:
+    #    _clear_grads_if_needed(module._fsdp_handles(module))
+    if module._has_params:
```
Sorry, let me clarify the `_clear_grads_if_needed()` call.

The idea is that we want to free the gradient memory as early as possible, so, given that this `_clear_grads_if_needed()` incurs minimal overhead (looping over gradients and checking if all are `None`), I inserted this call at every FSDP callsite.

In that way, we should prefer to keep it as is since the root can help free all possible gradient memory, which may be earlier than deferring to the submodule.

fegin replied:
Right now, the pre_hook is not an actual pre_state_dict_hook. So it will be called from the bottom-up. So the gradient won't be freed in time for a leaf module before the leaf module's post_state_dict_hook is called. Once the pre_state_dict_hook is actually implemented in `nn.Module`, I'll change it back.

fegin commented:
@pytorchbot merge
